### PR TITLE
Task 15 companies support

### PR DESCRIPTION
**Pull request by stanleyszysz**

### CODE REVIEW

File: src/main/java/pl/fc/invoicing/controllers/CompanyController.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    public ResponseEntity<CompanyDto> update(@PathVariable UUID id, @RequestBody CompanyDto updatedCompany) {
+        log.debug("Updating invoice by id: " + id);
+        return ResponseEntity.ok(companyService.update(id, updatedCompany));
+    }
+
+    @Override
+    public ResponseEntity<Void> delete(@PathVariable UUID id) {
+        try {
+            companyService.delete(id);
```
Comment (radoslawdabrowski) [inline src/main/java/pl/fc/invoicing/controllers/CompanyController.java:57]: I would suggest to change this implementation to use some ExceptionHandler

File: src/main/java/pl/fc/invoicing/controllers/InvoiceController.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Override
+    public ResponseEntity<InvoiceDto> update(@PathVariable UUID id, @RequestBody InvoiceDto updatedInvoice) {
+        log.debug("Updating invoice by id: " + id);
+        return ResponseEntity.ok(invoiceService.update(id, updatedInvoice));
+    }
+
+    @Override
+    public ResponseEntity<Void> delete(@PathVariable UUID id) {
+        try {
```
Comment (radoslawdabrowski) [inline src/main/java/pl/fc/invoicing/controllers/InvoiceController.java]: I would suggest to change this implementation to use some ExceptionHandler - in service just throw some implementation of RuntimeException and create ExceptionHandler with @ControllerAdvice

File: src/main/java/pl/fc/invoicing/dto/CompanyDto.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+public class CompanyDto {
+
+    private UUID companyId;
+    private String taxIdentifier;
+    private String name;
+    private Address address;
+    private BigDecimal pensionInsurance;
+    private BigDecimal healthInsurance;
+
+    public CompanyDto(Company company) {
```
Comment (radoslawdabrowski) [inline src/main/java/pl/fc/invoicing/dto/CompanyDto.java]: change constructor to some static method named for example of() - it is common case

File: src/main/java/pl/fc/invoicing/model/Company.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Builder.Default
+    @Schema(description = "Pension insurance amount", example = "1000.00", required = true)
+    private BigDecimal pensionInsurance = BigDecimal.ZERO;
+
+    @Builder.Default
+    @Schema(description = "Health insurance amount", example = "500.00", required = true)
+    private BigDecimal healthInsurance = BigDecimal.ZERO;
+
+    public Company(CompanyDto companyDto) {
```
Comment (radoslawdabrowski) [inline src/main/java/pl/fc/invoicing/model/Company.java]: change constructor to some static method named for example of()

File: src/main/java/pl/fc/invoicing/model/Invoice.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    @Schema(description = "Product name", required = true)
+    private List<InvoiceEntry> entries;
+
+    public void updateRelations() {
+        if (entries != null) {
+            entries.forEach(invoiceEntry -> invoiceEntry.setInvoice(this));
+        }
+    }
+
+    public void updateFromDto(InvoiceDto invoiceDto) {
```
Comment (radoslawdabrowski) [inline src/main/java/pl/fc/invoicing/model/Invoice.java]: maybe use Mapstruct

File: src/main/java/pl/fc/invoicing/services/JsonService.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
+package pl.fc.invoicing.services;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import org.springframework.stereotype.Service;
+
+@Service
+public class JsonService<T> {
```
Comment (radoslawdabrowski) [inline src/main/java/pl/fc/invoicing/services/JsonService.java:9]: is JsonService even needed?

File: src/main/java/pl/fc/invoicing/services/impl/InvoiceServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        invoiceModel.updateRelations();
+        Invoice savedInvoice = invoiceRepository.save(invoiceModel);
+        return modelMapper.map(savedInvoice, InvoiceDto.class);
+    }
+
+    @Override
+    public Optional<InvoiceDto> getById(UUID id) {
+        Optional<Invoice> invoice = invoiceRepository.findById(id);
+        if (invoice.isEmpty()) {
+            return Optional.empty();
```
Comment (radoslawdabrowski) [inline src/main/java/pl/fc/invoicing/services/impl/InvoiceServiceImpl.java]: you shouldn't return empty object. Just throw exception and handle it via ExceptionHandler

**Review by radoslawdabrowski — CHANGES_REQUESTED** [review]

File: src/main/java/pl/fc/invoicing/controllers/CompanyController.java (hunk 1 above)
Comment (stanleyszysz) [reply src/main/java/pl/fc/invoicing/controllers/CompanyController.java:57]: Corrected.

File: src/main/java/pl/fc/invoicing/controllers/InvoiceController.java (hunk 2 above)
Comment (stanleyszysz) [reply src/main/java/pl/fc/invoicing/controllers/InvoiceController.java]: Corrected.

File: src/main/java/pl/fc/invoicing/dto/CompanyDto.java (hunk 3 above)
Comment (stanleyszysz) [reply src/main/java/pl/fc/invoicing/dto/CompanyDto.java]: Corrected.

File: src/main/java/pl/fc/invoicing/model/Company.java (hunk 4 above)
Comment (stanleyszysz) [reply src/main/java/pl/fc/invoicing/model/Company.java]: Corrected.

File: src/main/java/pl/fc/invoicing/model/Invoice.java (hunk 5 above)
Comment (stanleyszysz) [reply src/main/java/pl/fc/invoicing/model/Invoice.java]: Corrected.

File: src/main/java/pl/fc/invoicing/services/impl/InvoiceServiceImpl.java (hunk 7 above)
Comment (stanleyszysz) [reply src/main/java/pl/fc/invoicing/services/impl/InvoiceServiceImpl.java]: Corrected.

File: src/main/java/pl/fc/invoicing/services/JsonService.java (hunk 6 above)
Comment (stanleyszysz) [reply src/main/java/pl/fc/invoicing/services/JsonService.java:9]: Json is needed for testing.

**Review by radoslawdabrowski — APPROVED** [review]

**Review by blazejbujala — APPROVED** [review]

**Review by radoslawdabrowski — APPROVED** [review]

**Pull request merged**